File: 1079-sum-of-root-to-leaf-binary-numbers/sum-of-root-to-leaf-binary-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathsum(self, root, num):
        if root == None:
            return 0

        cur = num * 2 + root.val
        if root.left == None and root.right == None:
            return cur

        left = self.pathsum(root.left, cur)
        right = self.pathsum(root.right, cur)
        return left + right

    def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:
        return self.pathsum(root, 0)

File: 1079-sum-of-root-to-leaf-binary-numbers/sum-of-root-to-leaf-binary-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pathsum(TreeNode* root, int num) {
        // encouter null, no value, early stopping
        if (root == NULL) {
            return 0;
        }

        int cur = (num << 1) + root->val;
        // encounter leaf node, return current number
        if (root->left == NULL && root->right == NULL) {
            return cur;
        }

        int left = pathsum(root->left, cur);
        int right = pathsum(root->right, cur);
        return left + right;
    }
    int sumRootToLeaf(TreeNode* root) {
        return pathsum(root, 0);
    }
};